### PR TITLE
Easier way to write controls

### PR DESCRIPTION
**Pull request by exyi**
Implemented `ValueOrBinding<T>` and `CompositeControl` types + various tweaks in the framework to make this possible. It's the first part of #442

`ValueOrBinding<T>` is a discriminated union of `IBinding` and `T`, with few helper methods. It's required when you want to carry DotvvmProperty in one variable - which is exactly what I wanted to do - put every dotvvm property into a simple function parameter.

`CompositeControl` is a control base class (formerly known as `EpicCoolControl`) It automatically creates and maps dotvvm properties from function parameters, without the need to declare them in the boilerplate way. The function simply gets the properties and returns dotvvm control tree which should be rendered - which implies that this way of rendering controls is not designed to replace the legacy one, as the returned control need to be rendered in some way ;). I think that each method has it's strong/weak spots and having both of them will help us cover more use-cases. Little example:

```C#
public class MyControl: CompositeControl {
     public static DotvvmControl GetContents(
          // this property is required and supports both bindings and hard-coded values
          ValueOrBinding<string> myProperty,
          // this one is optional and does not allow value binding (resource binding are allowed, it's simply evaluated before executing this method)
          string tagName = "div"
     )
     {
          return new HtmlGenericControl(tagName) {
             Children = { new Literal(myProperty) }
         };
   }
}
```

The `GetContents` method is found by name, it can be either static or instance and it has to return `DotvvmControl` or `IEnumerable<DotvvmControl>`. It's invoked `OnLoad` (only once, so evaluated bindings will not be updated, you should pass the evaluation to "primitive controls"). And it's still a DotvvmControl, so you can declare dotvvm properties in the old way, access attached or inherited properties, do something on lifecycle events and so on. 

It also supports all the attributes you'd expect from dotvvm property on the parameter - data context change, markup options and so on. And property groups are also supported, see [the sample](https://github.com/riganti/dotvvm/blob/easier-controls/src/DotVVM.Samples.Common/Controls/EpicCoolControlSample.cs)

### Control Capabilities

This PR also includes a feature called "control capabilities". It's essentially a group of properties that can be declared once and then reused. It's declared as a data class with the desired properties, see https://github.com/riganti/dotvvm/pull/535/files#diff-896c261fa17422d2cbad9aa5ee21140eR328 for an example. Then, when you want to support the capability, you just register a single property and pass the entire class to the "primitive control" or some helper function that will actually render it.

This feature is designed to replace the stupid antipattern of using inheritance to for only getting some common properties (see `ItemsControl`, ...)

### CODE REVIEW

**Comment by exyi** [comment]
Note: It will have to be supported in the Visual Studio extension - we will either have to dump properties from DotVVM.Compiler or add support for these controls. cc @Mylan719

**Comment by tomasherceg** [comment]
We've discussed an interesting thing with @quigamdev - what about the inheritance of composite controls? 

Naturally, the `GetContent` in the derived class will need to request all properties of the base class, and I think we should not shadow the base class properties - we should "de-duplicate" them, possibly by name.

```
public class Parent : CompositeControl {
  public static DotvvmControl GetContents(
    ValueOrBinding<string> prop1
  ) {
    ...
  }
}
public class Child : Parent {
  public static DotvvmControl GetContents(
    ValueOrBinding<string> prop1,
    ValueOrBinding<string> prop2
  ) {
 
    ...
       Parent.GetContents(prop1)   // or new Parent(prop1)
    ...

  }
}
```

I think that we should register prop1 only for the Parent class and not for the Child class (and throw an exception if there is a same-named property in the Child class that has a different type).

It would be very useful in Bootstrap for DotVVM. 
For example, we have plenty of controls that derive from [FormGroup](https://www.dotvvm.com/docs/controls/bootstrap4/FormGroup/2.0): `TextBoxFormGroup`, `ComboBoxFormGroup` etc. 
All of these controls have the `Label` property, and I think we have some other control in which we browse all form groups in the form, and need to access the labels universally.

**Comment by quigamdev** [comment]
Merging this PR to `alpha-testing` for further testing and revision of the design.

**Comment by Mylan719** [comment]
We seems to have en issue here with the newly added HtmlCapability property registered under the name "Html" on the HtmlGenericControl. The control HtmlLiteral is based on HtmlGenericControl and has own property named "Html".

File: src/DotVVM.Framework.Tests/ControlTests/CompositeControlTests.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+                <cc:WrappedHtmlControl TagName=div> Something here </cc:WrappedHtmlControl>
+                "
+            );
+
+            check.CheckString(r.FormattedHtml, fileExtension: "html");
+        }
+
+
+        // [TestMethod]
+        // public async Task CommandBinding()
```
Comment (quigamdev) [inline src/DotVVM.Framework.Tests/ControlTests/CompositeControlTests.cs:52]: Please remove the commented code

File: src/DotVVM.Framework/Binding/ValueOrBinding.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+        public ValueOrBinding<T2> UpCast<T2>()
+            where T2 : T =>
+            this.binding != null ?
+            new ValueOrBinding<T2>(this.binding) :
+            new ValueOrBinding<T2>((T2)this.value!);
+
+        public ParametrizedCode GetParametrizedJsExpression(DotvvmBindableObject control, bool unwrapped = false) =>
+            ProcessValueBinding(control,
+                value => new ParametrizedCode(JsonConvert.SerializeObject(value), OperatorPrecedence.Max),
```
Comment (exyi) [inline src/DotVVM.Framework/Binding/ValueOrBinding.cs:80]: Serialization option that we use elsewhere.

File: src/DotVVM.Framework/Compilation/ControlTree/ControlTreeResolverBase.cs [hunk 3]
```diff
@@ -652,21 +652,21 @@ public void ProcessControlContent(IAbstractControl control, IEnumerable<DothtmlN
                     content.Add(node);
                 }
             }
-            if (control.Metadata.DefaultContentProperty != null)
+            if (control.Metadata.DefaultContentProperty is IPropertyDescriptor contentProperty)
```
Comment (exyi) [inline src/DotVVM.Framework/Compilation/ControlTree/ControlTreeResolverBase.cs:655]:
```suggestion
            if (control.Metadata.DefaultContentProperty is {} contentProperty)
```

File: src/DotVVM.Framework/Controls/CompositeControl.cs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                var arguments = method.GetParameters().Select(initializeArgument);
+
+                if (!controlInfoCache.TryAdd(controlType, new ControlInfo { GetContentsMethod = method, Properties = arguments.ToImmutableArray() }))
+                    throw new Exception("no");
+            }
+        }
+
+        protected internal override void OnLoad(IDotvvmRequestContext context)
+        {
+            this.Children.Clear();
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/CompositeControl.cs:72]: Throw exception when not empty.

File: src/DotVVM.Framework/Controls/DotvvmBindableObject.cs [hunk 5]
```diff
@@ -85,16 +85,16 @@ public T GetValue<T>(DotvvmProperty property, bool inherit = true)
         internal object? EvalPropertyValue(DotvvmProperty property, object? value)
         {
             if (property.IsBindingProperty) return value;
-            if (value is IBinding)
+            while (value is IBinding binding) //TODO: MERGE CONFLICT
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/DotvvmBindableObject.cs:88]: This is badly merged

File: src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs [hunk 6]
```diff
@@ -22,21 +22,166 @@ public static class DotvvmBindableObjectHelper
...
             return DotvvmProperty.ResolveProperty(property.DeclaringType!, property.Name) ?? throw new Exception($"Property '{property.DeclaringType!.Name}.{property.Name}' is not a registered DotvvmProperty.");
         }
+        public static DotvvmProperty GetDotvvmProperty<TControl>(this TControl control, string propName)
+            where TControl : DotvvmBindableObject
+        {
+            return DotvvmProperty.ResolveProperty(typeof(TControl), propName) ?? throw new Exception($"Property '{typeof(TControl)}.{propName}' is not a registered DotvvmProperty.");
+        }
 
-        public static TControl SetBinding<TControl, TProperty>(this TControl control, Expression<Func<TControl, TProperty>> prop, IBinding? binding)
+        public static TControl SetProperty<TControl, TProperty>(this TControl control, Expression<Func<TControl, TProperty>> prop, IBinding? binding)
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs:31]: Leave the old SetBinding and SetValue function and make them obsolete.

File: src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs [hunk 7]
```diff
@@ -22,21 +22,166 @@ public static class DotvvmBindableObjectHelper
...
+            return control;
+        }
+
+        public static TControl SetAttribute<TControl, TProperty>(this TControl control, string attribute, ValueOrBinding<TProperty> value)
+            where TControl : IControlWithHtmlAttributes
+        {
+            return SetAttribute(control, attribute, value.BindingOrDefault ?? value.BoxedValue);
+        }
+
+        public static TControl SetAttribute<TControl, TProperty>(this TControl control, string attribute, ValueOrBinding<TProperty>? value)
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs:153]:
```suggestion
        public static TControl SetAttribute<TControl, TValue>(this TControl control, string attribute, ValueOrBinding<TValue>? value)
```

File: src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs [hunk 8]
```diff
@@ -22,21 +22,166 @@ public static class DotvvmBindableObjectHelper
...
+            {
+                var c = DotvvmCapabilityProperty.Find(typeof(TControl), typeof(TCapability), prefix);
+                if (c is null)
+                    throw new Exception($"Capability {prefix}{typeof(TCapability)} is not defined on {typeof(TControl)}");
+                c.SetValue(control, capability);
+            }
+            return control;
+        }
+
+        public static T AppendChildren<T>(this T control, IEnumerable<DotvvmControl>? children) where T : DotvvmControl
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs:172]: Params overload.

Comment (exyi) [reply src/DotVVM.Framework/Controls/DotvvmBindableObjectHelper.cs:172]: Allow nulls in the collection

File: src/DotVVM.Framework/Controls/HtmlGenericControl.cs [hunk 9]
```diff
@@ -99,6 +126,29 @@ public bool Visible
...
             DotvvmProperty.Register<bool, HtmlGenericControl>(t => t.Visible, true);
 
+
+        public HtmlCapability HtmlCapability
+        {
+            get => (HtmlCapability)this.GetValue(HtmlCapabilityProperty)!;
+            set => this.SetValue(HtmlCapabilityProperty, value);
+        }
+        public static readonly DotvvmCapabilityProperty HtmlCapabilityProperty =
+            DotvvmCapabilityProperty.RegisterCapability("HtmlCapability", typeof(HtmlGenericControl), typeof(HtmlCapability),
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/HtmlGenericControl.cs:136]: Add overload with generic argument like we have for DotvvmProperty

File: src/DotVVM.Framework/Controls/TextOrContentCapability.cs [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+            if (text is object && !"".Equals(text))
+                return new TextOrContentCapability { Text = ValueOrBinding<string>.FromBoxedValue(text) };
+            else
+                return new TextOrContentCapability { Content = control.Children.ToList() };
+        }
+
+        public void WriteToChildren(DotvvmControl control, DotvvmProperty textProperty)
+        {
+            control.SetProperty(textProperty, Text);
+            control.Children.Clear();
```
Comment (exyi) [inline src/DotVVM.Framework/Controls/TextOrContentCapability.cs:28]: Throw is not empty.

File: src/DotVVM.Samples.Common/Controls/CompositeControlSample.cs [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+            [CollectionElementDataContextChange(1)]
+            IValueBinding<string> titleBinding,
+
+            [ControlPropertyBindingDataContextChange("DataSource")]
+            [CollectionElementDataContextChange(1)]
+            IValueBinding<int?> numberBinding,
+
+            HtmlCapability html,
+
+            [DotvvmControlCapability(prefix: "inner-li:")]
```
Comment (exyi) [inline src/DotVVM.Samples.Common/Controls/CompositeControlSample.cs:27]: Make the : separator default

**Pull request merged**